### PR TITLE
python: optimize and unify PY2 and 3

### PR DESCRIPTION
**Pull request by methane**
- cache method before loop
- use dict instead of regular expression to categorizing chars
- unify Python 2 and 3 code
- add test

```
$ python benchmark.py
## benchmarker:         release 4.0.1 (for python)
## python version:      2.7.10
## python compiler:     GCC 4.2.1 Compatible Apple LLVM 6.1.0 (clang-602.0.53)
## python platform:     Darwin-14.5.0-x86_64-i386-64bit
## python executable:   /Users/inada-n/venvs/py2/bin/python
## cpu model:           Intel(R) Core(TM) i5-4278U CPU @ 2.60GHz
## parameters:          loop=10, cycle=1, extra=0

##                        real    (total    = user    + sys)
tokenize                7.5446    7.5300    7.5100    0.0200

## Ranking                real
tokenize                7.5446  (100.0) ********************

## Matrix                 real    [01]
[01] tokenize           7.5446   100.0

$ python3 benchmark.py
## benchmarker:         release 4.0.1 (for python)
## python version:      3.5.0
## python compiler:     GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.1.76)
## python platform:     Darwin-14.5.0-x86_64-i386-64bit
## python executable:   /Users/inada-n/pyenv/versions/3.5.0/bin/python3
## cpu model:           Intel(R) Core(TM) i5-4278U CPU @ 2.60GHz
## parameters:          loop=10, cycle=1, extra=0

##                        real    (total    = user    + sys)
tokenize                9.4400    9.4400    9.4000    0.0400

## Ranking                real
tokenize                9.4400  (100.0) ********************

## Matrix                 real    [01]
[01] tokenize           9.4400   100.0
```

### CODE REVIEW

**Comment by chezou** [comment]
Your mean following patch, are you https://github.com/chezou/TinySegmenter.jl/compare/08cc32c...bc76c45

File: benchmark/tinysegmenter3.py [hunk 1]
```diff
@@ -106,8 +102,53 @@ def tokenize(text):
     p1 = "U"
     p2 = "U"
     p3 = "U"
-    for i in range(4, len(seg) - 3):
 
+    # method cache
```
Comment (chezou) [inline benchmark/tinysegmenter3.py]: looks very Pythonic way :)

**Comment by chezou** [comment]
looks good :)

**Comment by chezou** [comment]
FYI @samuraiT

**Comment by chezou** [comment]
@methane  Can you rebase on master?

**Comment by methane** [comment]
@chezou done.

File: benchmark/test_tinysegmenter.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+
+def test_timemachine():
+    with io.open('../test/timemachineu8j.txt', encoding='utf-8') as f:
+        text = f.read()
+
+    toks = tinysegmenter.tokenize(text)
+
+    with io.open('tokenized.txt', 'w', encoding='utf-8') as f:
+        f.write(' | '.join(toks))
```
Comment (chezou) [inline benchmark/test_tinysegmenter.py]: Will this file be cleared after test?

**Comment by chezou** [comment]
Other modifications are looks good to me.

File: benchmark/test_tinysegmenter.py (hunk 2 above)
Comment (methane) [inline benchmark/test_tinysegmenter.py]: fixed

**Comment by chezou** [comment]
🎉

**Comment by chezou** [comment]
Thank you for contribution. I'll benchmark optimized code!

**Comment by chezou** [comment]
After optimization, Python 2 and 3 are same faster almost Node.js.

| JavaScript (Node.js) | Python2 | Python3 | Julia | Ruby |
| --- | --- | --- | --- | --- |
| 9.86 | 11.33 | 11.47 | 1.31 | 17.29 |

**Pull request merged**